packages/squiggle-lang/src/ast/serialize.ts

```diff
@@ -0,0 +1,323 @@
+        args: node.args.map(visit.ast),
+      };
+    case "InfixCall":
+      return {
+        ...node,
+        args: [visit.ast(node.args[0]), visit.ast(node.args[1])],
+      };
+    case "UnaryCall":
+      return {
+        ...node,
+        arg: visit.ast(node.arg),
+      };
+    case "Pipe":
+      return {
+        ...node,
+        leftArg: visit.ast(node.leftArg),
+        fn: visit.ast(node.fn),
+        rightArgs: node.rightArgs.map(visit.ast),
+      };
+    case "Decorator":
+      return {
+        ...node,
+        name: visit.ast(node.name),
+        args: node.args.map(visit.ast),
+      };
+    case "DotLookup":
+      return {
+        ...node,
+        arg: visit.ast(node.arg),
+      };
+    case "BracketLookup":
+      return {
+        ...node,
+        arg: visit.ast(node.arg),
+        key: visit.ast(node.key),
+      };
+    case "Ternary":
+      return {
+        ...node,
+        condition: visit.ast(node.condition),
+        trueExpression: visit.ast(node.trueExpression),
+        falseExpression: visit.ast(node.falseExpression),
+      };
+    case "Identifier":
+      return node;
+    case "IdentifierWithAnnotation":
+      return {
+        ...node,
+        annotation: visit.ast(node.annotation),
+      };
+    case "Float":
+    case "String":
+    case "Boolean":
+      return node;
+    default:
+      throw node satisfies never;
+  }
+}
+
+export function deserializeAstNode(
+  node: SerializedASTNode,
+  visit: SquiggleDeserializationVisitor
+): ASTNode {
+  switch (node.type) {
+    case "Program":
+      return {
+        ...node,
+        imports: node.imports.map((item) => [
+          visit.ast(item[0]) as TypedNode<"String">,
+          visit.ast(item[0]) as TypedNode<"Identifier">,
+        ]),
+        statements: node.statements.map(visit.ast),
+        symbols: Object.fromEntries(
+          Object.entries(node.symbols).map(([key, value]) => [
+            key,
+            visit.ast(value),
+          ])
+        ),
+      };
+    case "Block":
+      return {
+        ...node,
+        statements: node.statements.map(visit.ast),
+      };
+    case "DecoratedStatement":
+      return {
+        ...node,
+        decorator: visit.ast(node.decorator) as TypedNode<"Decorator">,
+        statement: visit.ast(node.statement) as TypedNode<"LetStatement">,
+      };
+    case "LetStatement":
+      return {
+        ...node,
+        variable: visit.ast(node.variable) as TypedNode<"Identifier">,
+        value: visit.ast(node.value),
+      };
+    case "DefunStatement":
+      return {
+        ...node,
+        variable: visit.ast(node.variable) as TypedNode<"Identifier">,
+        value: visit.ast(node.value) as NamedNodeLambda,
+      };
+    case "Lambda":
+      return {
+        ...node,
+        args: node.args.map(visit.ast),
+        body: visit.ast(node.body),
+      };
+    case "Array":
+      return {
+        ...node,
+        elements: node.elements.map(visit.ast),
+      };
```

Comment:
I could find a way to shorten this code, with helpers such as `serialize(node, { fields: [], arrayFields: ['elements'] })`, or even shape-guessing checks, but intentionally decided against this.

The reason is performance: JS heavily optimizes objects when all keys are static, and I expect any helper code like this would things slower.

Boilerplate is annoying, but thanks to TypeScript it's pretty robust: if we forget to serialize a field, TypeScript will complain, in most cases. One exception is nullable fields, but there aren't that many of those in AST.

It will also fail if we forget to serialize some specific type, thanks to `satisfies never` trick in `default:`.

The only "better" option that I can see here, at scale (if we had 100+ node types), is code generation. But it's obviously not worth it in this case.